### PR TITLE
Mongo enhancements

### PR DESCRIPTION
**Pull request by prateek-sumologic**

### CODE REVIEW

File: sumomongodbatlascollector/api.py [hunk 1]
```diff
@@ -20,21 +23,21 @@ class MongoDBAPI(BaseAPI):
     def __init__(self, kvstore, config):
         super(MongoDBAPI, self).__init__(kvstore, config)
         self.api_config = self.config['MongoDBAtlas']
+        self.MAX_REQUEST_WINDOW_LENGTH = self.api_config['Collection']['MAX_REQUEST_WINDOW_LENGTH'] if self.api_config['Collection']['MAX_REQUEST_WINDOW_LENGTH'] else 3600
```
Comment (himanshu219) [inline sumomongodbatlascollector/api.py:26]: we can replace it with single line `self.api_config['Collection'].get('MAX_REQUEST_WINDOW_LENGTH', 3600)`

File: sumomongodbatlascollector/api.py [hunk 2]
```diff
@@ -70,12 +78,13 @@ def fetch(self):
...
                     self.log.error(f'''Failed to send LogType: {self.get_key()}''')
             elif fetch_success and len(content) == 0:
-                self.log.info(f'''No results window LogType: {log_type} kwargs: {kwargs} status: {fetch_success}''')
+                self.log.info(f'''No results window LogType: {log_type} kwargs: {kwargs} status: {fetch_success} url: {url}''')
                 is_move_fetch_window, new_state = self.check_move_fetch_window(kwargs)
                 if is_move_fetch_window:
                     self.save_state(**new_state)
                     self.log.debug(f'''Moving fetched window newstate: {new_state}''')
             else:
+                self.log.info(f'''No results window LogType: {log_type} kwargs: {kwargs} status: {fetch_success} url: {url}''')
```
Comment (himanshu219) [inline sumomongodbatlascollector/api.py:87]: else condition is a error condition so we can put url and kwargs in self.log.error statement in line 88

File: sumomongodbatlascollector/api.py [hunk 3]
```diff
@@ -57,7 +66,6 @@ def fetch(self):
         state = None
         payload = []
         try:
-
             fetch_success, content = ClientMixin.make_request(url, method="get", logger=self.log, TIMEOUT=self.collection_config['TIMEOUT'], MAX_RETRY=self.collection_config['MAX_RETRY'], BACKOFF_FACTOR=self.collection_config['BACKOFF_FACTOR'], **kwargs)
```
Comment (himanshu219) [inline sumomongodbatlascollector/api.py:69]: can we add one more logline here as well with  
LogType: {log_type} kwargs: {kwargs} status: {fetch_success} url: {url}
execution time: {end_time - start_time:.2f} seconds current_process_memory_usage: {self._get_current_process_memory_usage()}

Comment (himanshu219) [reply sumomongodbatlascollector/api.py:69]: Also at line 77 and 79 also lets add kwargs url execution time and current_process_memory_usage

File: sumomongodbatlascollector/api.py [hunk 4]
```diff
@@ -92,11 +101,14 @@ def fetch(self):
...
         sess = ClientMixin.get_new_session()
-        self.log.info(f'''Fetching LogType: {log_type}  starttime: {kwargs['params']['minDate']} endtime: {kwargs['params']['maxDate']}''')
+        self.log.info(f'''Fetching LogType: {log_type}  starttime: {kwargs['params']['minDate']} endtime: {kwargs['params']['maxDate']} url: {url}''')
         try:
             while next_request:
                 send_success = has_next_page = False
+                start_time = time.time()
                 status, data = ClientMixin.make_request(url, method="get", session=sess, logger=self.log, TIMEOUT=self.collection_config['TIMEOUT'], MAX_RETRY=self.collection_config['MAX_RETRY'], BACKOFF_FACTOR=self.collection_config['BACKOFF_FACTOR'], **kwargs)
+                end_time = time.time()
+                self.log.info(f'''Fetching LogType: {log_type}  starttime: {kwargs['params']['minDate']} endtime: {kwargs['params']['maxDate']} url: {url} execution time: {end_time - start_time:.2f} seconds current_process_memory_usage: {self._get_current_process_memory_usage()}''')
```
Comment (himanshu219) [inline sumomongodbatlascollector/api.py:111]: lets add current_process_memory_usage execution time for line 121 and 135 basically to track how much time it took to send to sumo

File: sumomongodbatlascollector/main.py [hunk 5]
```diff
@@ -209,18 +228,45 @@ def run(self):
...
             if not self.is_process_running(["sumomongodbatlascollector"]):
                 self.kvstore.release_lock_on_expired_key(self.SINGLE_PROCESS_LOCK_KEY, expiry_min=10)
 
+    def execute_api_with_logging(self, apiobj):
+        api_type = str(apiobj.__class__.__name__)
+        start_time = time.time()
+        start_memory = tracemalloc.get_traced_memory()[0]
+        try:
+            result = apiobj.fetch()
+        except Exception as exc:
```
Comment (himanshu219) [inline sumomongodbatlascollector/main.py:250]: not needed we can remove this exeception

**Review by himanshu219 — CHANGES_REQUESTED** [review]
would be good to create a context manager for logging execution time and memory so that we can disable it if we want.

File: sumomongodbatlascollector/api.py (hunk 1 above)
Comment (prateek-sumologic) [reply sumomongodbatlascollector/api.py]: Have changes this to a .get lookup with default values.

File: sumomongodbatlascollector/main.py (hunk 5 above)
Comment (prateek-sumologic) [reply sumomongodbatlascollector/main.py]: Removed this exception,now we are doing time and process memory logging with context manager

**Comment by prateek-sumologic** [comment]
> would be good to create a context manager for logging execution time and memory so that we can disable it if we want.

Have added a new context manager

File: requirements.txt [hunk 6]
```diff
@@ -1 +1,2 @@
-sumologic-appclient-sdk==1.0.17
+sumologic-appclient-sdk==2.0.10
+requests>=2.32.2
```
Comment (npande) [inline requirements.txt:2]: Upgrade to 2.32.3

File: setup.py [hunk 7]
```diff
@@ -26,9 +26,9 @@
     packages=find_packages(),
     install_requires=INSTALL_REQUIRES,
     extras_require={
-        'aws': ["boto3>=1.9.66", "botocore>=1.12.66"],
-        "gcp": ["google-api-python-client>=1.7.8", "oauth2client>=4.1.3", "google-cloud-datastore>=1.7.3"],
-        "azure": ["azure-cosmosdb-table>=1.0.5", "bson>=0.5.8"]
+        'aws': ["boto3>=1.34.110", "botocore>=1.34.100"],
```
Comment (npande) [inline setup.py:29]: See if we can upgrade to latest 
boto3 1.34.151
botocore 1.34.151
google-api-python-client 2.139.0

File: sumomongodbatlascollector/api.py (hunk 2 above)
Comment (prateek-sumologic) [reply sumomongodbatlascollector/api.py]: Moved to the context manager based logging

File: sumomongodbatlascollector/api.py (hunk 3 above)
Comment (prateek-sumologic) [reply sumomongodbatlascollector/api.py]: Moved to the context manager based logging

File: sumomongodbatlascollector/api.py (hunk 4 above)
Comment (prateek-sumologic) [reply sumomongodbatlascollector/api.py]: Moved to the context manager based logging

File: sumomongodbatlascollector/main.py [hunk 8]
```diff
@@ -149,6 +164,9 @@ def build_task_params(self):
         process_ids, hostnames = self._get_process_names()
         cluster_mapping = self.kvstore.get("cluster_mapping", {}).get("values", {})
 
+
+        start_time = time.time()
+        self.log.debug(f"API level logging started at {start_time}, memory usage: {tracemalloc.get_traced_memory()[0]} bytes")
```
[the remark this answers is not in the document]
Comment (prateek-sumologic) [reply sumomongodbatlascollector/main.py:170]: Moved to the context manager based logging

File: template.yaml [hunk 9]
```diff
@@ -68,9 +71,32 @@ Resources:
...
+          - dynamodb:Query
+          Resource: !GetAtt MongoDBAtlasTableResource.Arn
       Runtime: python3.11
       Timeout: 900
     Type: AWS::Serverless::Function
+
+	MongoDBAtlasTableResource:
+		Type: AWS::DynamoDB::Table
+		Properties:
+			TableName: !Sub ${AWS::StackName}-mongodbatlastable
```
Comment (himanshu219) [inline template.yaml:92]: need to pass DBNAME as env variable

File: template.yaml [hunk 10]
```diff
@@ -68,9 +71,32 @@ Resources:
...
       MemorySize: 256
-      Policies:
-      - AmazonDynamoDBFullAccess
+			Policies:
+			- Version: "2012-10-17"
+				Statement:
+				- Effect: Allow
+					Action:
+					- dynamodb:PutItem
+          - dynamodb:UpdateItem
```
Comment (himanshu219) [inline template.yaml:80]: use DynamoDBWritePolicy and DynamoDBReadPolicy

File: template.yaml (hunk 9 above)
Comment (himanshu219) [reply template.yaml:92]: use stack id as a suffix

File: template.yaml [hunk 11]
```diff
@@ -68,9 +71,32 @@ Resources:
...
+
+	MongoDBAtlasTableResource:
+		Type: AWS::DynamoDB::Table
+		Properties:
+			TableName: !Sub ${AWS::StackName}-mongodbatlastable
+			AttributeDefinitions:
+			- AttributeName: id
+				AttributeType: S
+			KeySchema:
+			- AttributeName: id
```
Comment (himanshu219) [inline template.yaml:97]: attribute definition should be same as the current table https://github.com/SumoLogic/sumologic-appclient-sdk/blob/f47a7b94ab5fa071d4a07af639382742008c28f8/sumoappclient/omnistorage/aws.py#L241

read write capacities should also be same as current configuration https://github.com/SumoLogic/sumologic-appclient-sdk/blob/f47a7b94ab5fa071d4a07af639382742008c28f8/sumoappclient/omnistorage/aws.py#L256

File: sumomongodbatlascollector/api.py [hunk 12]
```diff
@@ -2,39 +2,58 @@
```
Comment (himanshu219) [inline sumomongodbatlascollector/api.py:2]: may need to replace with utf-8

File: requirements.txt [hunk 13]
```diff
@@ -1 +1,3 @@
-sumologic-appclient-sdk==1.0.17
+sumologic-appclient-sdk==2.0.10
+requests>=2.32.3
+future-fstrings
```
Comment (himanshu219) [inline requirements.txt:3]: do we need future-fstrings since now > 3 it already supports

**Comment by icco** [comment]
Just checking on the status of this. We're already running this PR in production as the existing version just doesn't work at all, would love to see this merged and officially released.

**Comment by himanshu219** [comment]
Hi @icco , we were not able to test this in an environment with heavy load of logs, would you be able to help us with below queries
1. How much throughput of logs are you able to receive in sumo logic ? 
2. Are you still seeing any delay in log ingestion? 
3. Also did you tried filtering the clusters  using `Clusters: ["cluster1-shard-00-00.abc123.mongodb.net"]` parameter ?
4. Did you see any reduction in Lambda timeout and OOM errors ?

**Comment by icco** [comment]
> Hi @icco , we were not able to test this in an environment with heavy load of logs, would you be able to help us with below queries
> 
> 1. How much throughput of logs are you able to receive in sumo logic ?

Looking at Sumo's reporting, it looks like 2Gi per day on average. If you've got a specific query you want me to run, I'd be happy to.

> 2. Are you still seeing any delay in log ingestion?

5 - 10 minute delay, but this is due to our implementation. 

> 3. Also did you tried filtering the clusters  using `Clusters: ["cluster1-shard-00-00.abc123.mongodb.net"]` parameter ?

We have not.

> 4. Did you see any reduction in Lambda timeout and OOM errors ?

Yes, although we don't deploy this in Lambda, we deploy this in kubernetes.

**Comment by himanshu219** [comment]
Thanks for the info, we will release it by End of this month.

**Review by himanshu219 — APPROVED** [review]

**Pull request merged**